### PR TITLE
trigger preprocess and train script in local mode

### PR DESCRIPTION
**Pull request by joker2411**
Changes to be committed:
	modified:   src/ml_core/preprocess_and_train.py
	modified:   src/processors/AWSProcessor.py
	modified:   src/processors/LocalProcessor.py
	modified:   src/processors/Processor.py
	modified:   src/processors/SnowflakeProcessor.py

## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/processors/Processor.py [hunk 1]
```diff
@@ -23,24 +21,6 @@ def __init__(
         self.connector = connector
         self.session = session
 
-    def train(
```
Comment (dpatchigolla) [inline src/processors/Processor.py:26]: Why not leave it as an abstractmethod?

File: src/ml_core/preprocess_and_train.py [hunk 2]
```diff
@@ -310,7 +311,11 @@ def preprocess_and_train(
         trainer = RegressionTrainer(**args.merged_config)
 
     # Creating the Redshift connector and session bcoz this case of code will only be triggerred for Redshift
-    current_dir = os.path.dirname(os.path.abspath(__file__))
+    current_dir = (
```
Comment (dpatchigolla) [inline src/ml_core/preprocess_and_train.py:314]: Why is this change required?

File: src/ml_core/preprocess_and_train.py [hunk 3]
```diff
@@ -293,6 +293,7 @@ def preprocess_and_train(
     parser.add_argument("--merged_config", type=json.loads)
     parser.add_argument("--prediction_task", type=str)
     parser.add_argument("--wh_creds", type=json.loads)
+    parser.add_argument("--output_path", type=str)
```
Comment (dpatchigolla) [inline src/ml_core/preprocess_and_train.py:296]: A new parameter got added but it is not used when calling in awsprocessor. how does it work?

File: src/ml_core/preprocess_and_train.py [hunk 4]
```diff
@@ -334,31 +339,32 @@ def preprocess_and_train(
...
             args.s3_bucket,
             args.aws_region_name,
             args.s3_path,
             local_folder,
             train_upload_whitelist,
         )
-    logger.debug(f"Deleting additional local directory from infra mode")
-    connector.cleanup(delete_local_data=True)
+        logger.debug(f"Deleting additional local directory from infra mode")
+        connector.cleanup(delete_local_data=True)
```
Comment (dpatchigolla) [inline src/ml_core/preprocess_and_train.py:370]: Can the logger statements and cleanup command be moved outside the if else blocks?

Comment (dpatchigolla) [reply src/ml_core/preprocess_and_train.py:370]: Or can we add mode as a param to upload_directory itself, and merge both functions to one?  This would prevent the need for an if else statment.
Within upload_directory, it attempts to fetch keys. If it finds, it sets up s3 client with those keys. else, without - similar to how it is currently in upload_directory. This may need changes in other places where these functions are currently being called.

**Review by dpatchigolla — CHANGES_REQUESTED** [review]

File: src/ml_core/preprocess_and_train.py (hunk 2 above)
Comment (joker2411) [reply src/ml_core/preprocess_and_train.py:314]: in local processing mode, we'll get the pb_output folder path which we'll use to create connector. But on ec2/k8, that path may or maynot exist because of churn repo not being there. Therefore, on ec2/k8, I kept it relative to project path for temporary use and processing.

File: src/ml_core/preprocess_and_train.py (hunk 3 above)
Comment (joker2411) [reply src/ml_core/preprocess_and_train.py:296]: because in local_mode, output path will be to pb output folder and in case of awsprocessor, we are using current project path. Also, I have checked that if we don't pass any argument, then ```parser.add_argument()``` has a default of None which works fine in our case.

File: src/ml_core/preprocess_and_train.py (hunk 4 above)
Comment (joker2411) [reply src/ml_core/preprocess_and_train.py:370]: Yes, I am also planning to move few things in __main__ into functions itself using mode as an additional parameter. Though I'll be taking that as a separate task after this and don't want to include that in this.

File: src/processors/Processor.py (hunk 1 above)
Comment (joker2411) [reply src/processors/Processor.py:26]: corrected.

**Review by dpatchigolla — APPROVED** [review]
